### DISCUSSION TITLE
typing for fetchall()

### DISCUSSION THREAD

**Discussion by davetapley** [duckdb/duckdb#7771] [General]
This:

```py
rel = con.sql("SELECT ts,val FROM data")
rows = rel.fetchall()
```

Gives ⬇️ in Pyright.

```
Type of "rows" is partially unknown
    Type of "rows" is "list[Unknown]" (reportUnknownVariableType)
```

---

I can `typing.cast` my way around it ⬇️  but I wonder if there is a more elegant way?

```py
rows = cast(list[tuple[datetime, int]], rel.fetchall())
```

---

It would be nice to give the type to `sql`, so it's with the SQL query which defines it.

Also I'm not sure if/how this relates to [`DuckDBPyType`](https://github.com/duckdb/duckdb-web/blob/579e5aeda85535770c52df76cddf420f2d45c3e7/docs/api/python/types.md)?

**Comment by Tishj** [comment]
That would likely require some dynamic typing for `fetchall`.
Nothing else has dynamic typing so I'm not even sure if that's possible within the package we use to create python bindings, but that could be worth looking into.
